src/cpp/src/multinomial_decoding.cpp

```diff
@@ -0,0 +1,256 @@
+// Copyright (C) 2023-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <numeric>
+#include <random>
+#include <regex>
+#include <vector>
+
+#include "openvino/genai/llm_pipeline.hpp"
+#include "utils.hpp"
+
+
+namespace {
+
+struct TokenIdScore {
+    int64_t id;
+    float score;
+
+    bool operator<(const TokenIdScore& other) const {
+        return score < other.score;
+    }
+
+    bool operator>(const TokenIdScore& other) const {
+        return score > other.score;
+    }
+};
+
+void apply_softmax_inplace(std::vector<TokenIdScore>& tokens) {
+    float max_score = std::max_element(tokens.begin(), tokens.end())->score;
+    float sum = 0.f;
+
+    for (auto& token : tokens) {
+        float s = std::exp(token.score - max_score);
+        token.score = s;
+        sum += s;
+    }
+
+    float inv_sum = 1.f / sum;
+
+    for (auto& token : tokens) {
+        token.score *= inv_sum;
+    }
+}
+
+TokenIdScore* sample_top_p(TokenIdScore* first, TokenIdScore* last, float top_p) {
+    // sort score
+    std::sort(first, last, std::greater<TokenIdScore>());
+
+    int tokens_size = last - first;
+    std::vector<TokenIdScore> token_scores(tokens_size);
+    for (size_t i = 0; i < tokens_size; i++) {
+        token_scores[i] = first[i];
+    }
+
+    // calculate softmax
+    apply_softmax_inplace(token_scores);
+
+    float prefix_sum = 0.0f;
+
+    // top_p
+    for (size_t i = 0; i < tokens_size; i++) {
+        prefix_sum += token_scores[i].score;
+        if (prefix_sum >= top_p) {
+            return first + (i + 1);
+        }
+    }
+
+    return last;
+}
+
+void apply_repetition_penalty(float* first, float* last, const std::vector<int64_t>& input_ids, float penalty) {
+    const float inv_penalty = 1.f / penalty;
+    const int vocab_size = last - first;
+    std::vector<bool> occurrence(vocab_size, false);
+    for (const int64_t id : input_ids) {
+        if (!occurrence[id]) {
+            first[id] *= (first[id] > 0) ? inv_penalty : penalty;
+        }
+        occurrence[id] = true;
+    }
+}
+
+void apply_inv_temperature(float* first, float* last, float inv_temperature) {
+    for (float* it = first; it != last; it++) {
+        *it *= inv_temperature;
+    }
+}
+
+struct RandomSampling {
+    const size_t top_k;
+    const float top_p;
+    const float inv_temperature;
+    const float repetition_penalty;
+
+    std::mt19937 gen{std::random_device{}()};
+
+    RandomSampling(ov::genai::GenerationConfig generation_config)
+        : top_k{generation_config.top_k},
+          top_p{generation_config.top_p},
+          inv_temperature{1.f / generation_config.temperature},
+          repetition_penalty{generation_config.repetition_penalty} {
+        generation_config.validate();

```

Comment:
@pavel-esir 
should we call validate in common part? Currently, each pipeline is responsible for it

Reply:
done